GenAPI incorrectly generates collection-based interface implementation

When targeting System.Collections.Immutable.1.7.1, GenAPI generates a class implementation which doesn't implement the interface that its member explicitly attempts to implement.

This illustrates the issue:

```c#
public sealed partial class ImmutableHashSet<T> : IEnumerable, ICollection
{
    bool Generic.ICollection<T>.IsReadOnly { get { throw null; } }
}
```

This results in a build error:

```
error CS0540: 'ImmutableHashSet<T>.ICollection<T>.IsReadOnly': containing type does not implement interface 'ICollection<T>'
```


You can see that the generated class doesn't implement any of the generic interfaces that it should:

```c#
public sealed class ImmutableHashSet<T> : IImmutableSet<T>, IReadOnlyCollection<T>, IEnumerable<T>, IEnumerable, IHashKeyCollection<T>, ICollection<T>, ISet<T>, ICollection, IStrongEnumerable<T, ImmutableHashSet<T>.Enumerator>
```

I am seeing this issue again.  Perhaps it is a slightly different scenario.

Steps to reproduce in https://github.com/dotnet/source-build-reference-packages:

1. ./generate.sh -x -p System.Security.Cryptography.X509Certificates,4.3.0
2. ./build.sh -sb

Build errors from generated code:

```
  /repos/source-build-reference-packages/artifacts/source-build/self/src/src/referencePackages/src/system.security.cryptography.x509certificates/4.3.0/ref/netstandard1.4/System.Security.Cryptography.X509Certificates.cs(339,34): error CS0539: 'X509CertificateCollection.Item' in explicit interface declaration is not found among members of the interface that can be implemented [/repos/source-build-reference-packages/artifacts/source-build/self/src/src/referencePackages/src/system.security.cryptography.x509certificates/4.3.0/System.Security.Cryptography.X509Certificates.4.3.0.csproj::TargetFramework=netstandard1.4] [/root/.nuget/packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23371.1/tools/Build.proj]
  /repos/source-build-reference-packages/artifacts/source-build/self/src/src/referencePackages/src/system.security.cryptography.x509certificates/4.3.0/ref/netstandard1.4/System.Security.Cryptography.X509Certificates.cs(373,34): error CS0683: 'X509CertificateCollection.get_Item(int)' explicit method implementation cannot implement 'IList.this[int].get' because it is an accessor [/repos/source-build-reference-packages/artifacts/source-build/self/src/src/referencePackages/src/system.security.cryptography.x509certificates/4.3.0/System.Security.Cryptography.X509Certificates.4.3.0.csproj::TargetFramework=netstandard1.4] [/root/.nuget/packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23371.1/tools/Build.proj]
  /repos/source-build-reference-packages/artifacts/source-build/self/src/src/referencePackages/src/system.security.cryptography.x509certificates/4.3.0/ref/netstandard1.4/System.Security.Cryptography.X509Certificates.cs(381,32): error CS0683: 'X509CertificateCollection.set_Item(int, object)' explicit method implementation cannot implement 'IList.this[int].set' because it is an accessor [/repos/source-build-reference-packages/artifacts/source-build/self/src/src/referencePackages/src/system.security.cryptography.x509certificates/4.3.0/System.Security.Cryptography.X509Certificates.4.3.0.csproj::TargetFramework=netstandard1.4] [/root/.nuget/packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23371.1/tools/Build.proj]
  /repos/source-build-reference-packages/artifacts/source-build/self/src/src/referencePackages/src/system.security.cryptography.x509certificates/4.3.0/ref/netstandard1.4/System.Security.Cryptography.X509Certificates.cs(319,104): error CS0738: 'X509CertificateCollection' does not implement interface member 'IList.this[int]'. 'X509CertificateCollection.this[int]' cannot implement 'IList.this[int]' because it does not have the matching return type of 'object'. [/repos/source-build-reference-packages/artifacts/source-build/self/src/src/referencePackages/src/system.security.cryptography.x509certificates/4.3.0/System.Security.Cryptography.X509Certificates.4.3.0.csproj::TargetFramework=netstandard1.4] [/root/.nuget/packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23371.1/tools/Build.proj]
```

The System.Collections.IList interface is not being generated correctly

Generated Code:

``` csharp
object Collections.IList.Item { get { throw null; } set { } }
object Collections.IList.get_Item(int index) { throw null; }
void Collections.IList.set_Item(int index, object value) { }
```

Expected Code:

``` csharp
object System.Collections.IList.this[int index] { get { throw null; } set { } }
```

cc @ericstj, @ViktorHofer, @andriipatsula

--- a/src/GenAPI/Microsoft.DotNet.GenAPI/Filtering/ImplicitSymbolFilter.cs
+++ b/src/GenAPI/Microsoft.DotNet.GenAPI/Filtering/ImplicitSymbolFilter.cs
@@ -28,7 +28,17 @@ namespace Microsoft.DotNet.GenAPI.Filtering
                     method.MethodKind == MethodKind.EventRemove ||
                     method.MethodKind == MethodKind.EventRaise ||
                     method.MethodKind == MethodKind.DelegateInvoke)
+                {
                     return false;
+                }
+
+                // If the method is an explicitly implemented getter or setter, exclude it.
+                // https://github.com/dotnet/roslyn/issues/53911
+                if (method.MethodKind == MethodKind.ExplicitInterfaceImplementation && 
+                    method.ExplicitInterfaceImplementations.Any(m => m is { MethodKind: MethodKind.PropertyGet or MethodKind.PropertySet }))
+                {
+                    return false;
+                }
             }
 
             if (symbol is ITypeSymbol type)

--- a/src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs
+++ b/src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs
@@ -89,6 +89,16 @@ namespace Microsoft.DotNet.GenAPI
                 }
             }
 
+            if (symbol is IPropertySymbol propertySymbol)
+            {
+                // Explicitly implemented indexers do not set IsIndexer
+                // https://github.com/dotnet/roslyn/issues/53911
+                if (!propertySymbol.IsIndexer && propertySymbol.ExplicitInterfaceImplementations.Any(i => i.IsIndexer))
+                {
+                    return syntaxGenerator.IndexerDeclaration(propertySymbol);
+                }
+            }
+
             try
             {
                 return syntaxGenerator.Declaration(symbol);

--- a/src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs
+++ b/src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs
@@ -2499,5 +2499,122 @@ namespace Microsoft.DotNet.GenAPI.Tests
                 }
                 """);
         }
+
+        [Fact]
+        public void TestExplicitInterfaceIndexer()
+        {
+            RunTest(original: """
+                    namespace a
+                    {
+                        public interface IFooList
+                        {
+                             object this[int index] { get; set; }
+                        }
+
+                        public struct Bar : IFooList
+                        {
+                        #pragma warning disable CS8597
+                            public string this[int index] { get { throw null; } set { } }
+                            object IFooList.this[int index] { get { throw null; } set { } }
+                        #pragma warning restore CS8597
+                        }
+                    }
+                    """,
+                expected: """
+                    namespace a
+                    {
+                        public partial struct Bar : IFooList
+                        {
+                            object IFooList.this[int index] { get { throw null; } set { } }
+                            public string this[int index] { get { throw null; } set { } }
+                        }
+
+                        public partial interface IFooList
+                        {
+                            object this[int index] { get; set; }
+                        }
+                    }
+                    """,
+                includeInternalSymbols: false);
+        }
+
+        [Fact]
+        public void TestExplicitInterfaceNonGenericCollections()
+        {
+            RunTest(original: """
+                    #nullable disable
+                    using System;
+                    using System.Collections;
+                    namespace a
+                    {
+                        #pragma warning disable CS8597
+                        
+                        public partial class MyStringCollection : ICollection, IEnumerable, IList
+                        {
+                            public int Count { get { throw null; } }
+                            public string this[int index] { get { throw null; } set { } }
+                            bool ICollection.IsSynchronized { get { throw null; } }
+                            object ICollection.SyncRoot { get { throw null; } }
+                            bool IList.IsFixedSize { get { throw null; } }
+                            bool IList.IsReadOnly { get { throw null; } }
+                            object IList.this[int index] { get { throw null; } set { } }
+                            public int Add(string value) { throw null; }
+                            public void AddRange(string[] value) { }
+                            public void AddRange(MyStringCollection value) { }
+                            public void Clear() { }
+                            public bool Contains(string value) { throw null; }
+                            public void CopyTo(string[] array, int index) { }
+                            public override int GetHashCode() { throw null; }
+                            public int IndexOf(string value) { throw null; }
+                            public void Insert(int index, string value) { }
+                            public void Remove(string value) { }
+                            public void RemoveAt(int index) { }
+                            void ICollection.CopyTo(Array array, int index) { }
+                            IEnumerator IEnumerable.GetEnumerator() { throw null; }
+                            int IList.Add(object value) { throw null; }
+                            bool IList.Contains(object value) { throw null; }                            
+                            int IList.IndexOf(object value) { throw null; }
+                            void IList.Insert(int index, object value) { }
+                            void IList.Remove(object value) { }
+                        }
+
+                        #pragma warning restore CS8597
+                    }
+                    """,
+                expected: """                    
+                    namespace a
+                    {
+                        public partial class MyStringCollection : System.Collections.ICollection, System.Collections.IEnumerable, System.Collections.IList
+                        {
+                            public int Count { get { throw null; } }
+                            public string this[int index] { get { throw null; } set { } }
+                            bool System.Collections.ICollection.IsSynchronized { get { throw null; } }
+                            object System.Collections.ICollection.SyncRoot { get { throw null; } }
+                            bool System.Collections.IList.IsFixedSize { get { throw null; } }
+                            bool System.Collections.IList.IsReadOnly { get { throw null; } }
+                            object System.Collections.IList.this[int index] { get { throw null; } set { } }
+                            public int Add(string value) { throw null; }
+                            public void AddRange(MyStringCollection value) { }
+                            public void AddRange(string[] value) { }
+                            public void Clear() { }
+                            public bool Contains(string value) { throw null; }
+                            public void CopyTo(string[] array, int index) { }
+                            public override int GetHashCode() { throw null; }
+                            public int IndexOf(string value) { throw null; }
+                            public void Insert(int index, string value) { }
+                            public void Remove(string value) { }
+                            public void RemoveAt(int index) { }
+                            void System.Collections.ICollection.CopyTo(System.Array array, int index) { }
+                            System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() { throw null; }
+                            int System.Collections.IList.Add(object value) { throw null; }
+                            bool System.Collections.IList.Contains(object value) { throw null; }
+                            int System.Collections.IList.IndexOf(object value) { throw null; }
+                            void System.Collections.IList.Insert(int index, object value) { }
+                            void System.Collections.IList.Remove(object value) { }
+                        }
+                    }
+                    """,
+                includeInternalSymbols: false);
+        }
     }
 }

